main.go

```diff
@@ -18,43 +19,41 @@ const (
 	agentSystemName  = "k3s-agent"
 )
 
-type K3sConfig struct {
-	ClusterInit bool     `yaml:"cluster-init,omitempty"`
-	Token       string   `yaml:"token,omitempty"`
-	Server      string   `yaml:"server,omitempty"`
-	TLSSan      []string `yaml:"tls-san,omitempty"`
-}
-
 func clusterProvider(cluster clusterplugin.Cluster) yip.YipConfig {
-	k3sConfig := K3sConfig{
+	k3sConfig := api.K3sConfig{
 		Token: cluster.ClusterToken,
 	}
 
+	var userOptionConfig string
 	switch cluster.Role {
 	case clusterplugin.RoleInit:
 		k3sConfig.ClusterInit = true
 		k3sConfig.TLSSan = []string{cluster.ControlPlaneHost}
+		userOptionConfig = cluster.Options
 	case clusterplugin.RoleControlPlane:
 		k3sConfig.Server = fmt.Sprintf("https://%s:6443", cluster.ControlPlaneHost)
 		k3sConfig.TLSSan = []string{cluster.ControlPlaneHost}
+		userOptionConfig = cluster.Options
 	case clusterplugin.RoleWorker:
 		k3sConfig.Server = fmt.Sprintf("https://%s:6443", cluster.ControlPlaneHost)
+		//Data received from upstream contains config for both control plane and worker. Thus, for worker, config is being filtered
+		//via unmarshal into agent config.
+		var agentCfg api.K3sAgentConfig
+		if err := yaml.Unmarshal([]byte(cluster.Options), &agentCfg); err == nil {
+			out, _ := yaml.Marshal(agentCfg)
+			userOptionConfig = string(out)
+		}
 	}
 
 	systemName := serverSystemName
 	if cluster.Role == clusterplugin.RoleWorker {
 		systemName = agentSystemName
 	}
 
-	// ensure we always have  a valid user config
-	if cluster.Options == "" {
```

Comment:
this should be moved above the switch statement, this implementation makes it possible for control plane and init nodes to have invalid configuration.

Reply:
@paynejacob If cluster.option is `yaml` then we don't need to initialize it. Earlier change of initializing cluster.option with `{}` was invalid as we were initializing it with empty json.